File: Problems/0600_0699/0606_Construct_String_from_Binary_Tree/Project_Scala/OperateTreeNode.scala
object OperateTreeNode {
    def createTreeNode(flds:String):TreeNode = {
        return createSubTreeNode(flds.split(","), 0, 0)
    }

    def createSubTreeNode(flds:Array[String], depth:Int, pos:Int):TreeNode = {
        if (flds.size <= 0)
            return null

        var cur_pos:Int = 0
        for (i <- 0 until depth)
            cur_pos = cur_pos + Math.pow(2, i).toInt
        
        if (cur_pos + pos > flds.size - 1)
            return null
        
        if (flds(cur_pos + pos) == "null")
            return null

        try {
            var node:TreeNode = new TreeNode(flds(cur_pos + pos).toInt)
            node.left = createSubTreeNode(flds, depth + 1, 2*pos)
            node.right = createSubTreeNode(flds, depth + 1, 2*pos + 1)

            return node

        } catch {
            case e: Exception =>

            println("\n" +  e + "\n" + "createTreeNode() Error ... flds(" + (cur_pos + pos).toString + ") = " + flds(cur_pos + pos) + "\n");
            sys.exit()
        }
    }

    var resultList:Array[String] = Array.empty

    def treeToStaircaseString(node:TreeNode):String = {
        var resultStr:String = treeToStaircaseSubString(node, 0)
        resultList = Array.empty;

        return resultStr;
    }

    def treeToStaircaseSubString(node:TreeNode, n:Int):String = {
        if (node == null)
            return ""
        
        if (resultList.size <= n) {
            resultList = resultList :+ "(" + node.value.toString() + ")"
        } else
            resultList(n) = resultList(n) + ",(" + node.value.toString() + ")"

        if (node.left != null)
            treeToStaircaseSubString(node.left, n + 1)
        if (node.right != null)
            treeToStaircaseSubString(node.right, n + 1)

        return resultList.mkString("\n") + "\n";
    }

    def tree2str(node:TreeNode):String = {
        if (node == null)
            return ""

        var resultStr:String = node.value.toString()

        if (node.left == null && node.right == null)
            return resultStr

        resultStr = resultStr + "(" + tree2str(node.left) + ")"
        if (node.right != null)
            resultStr = resultStr + "(" + tree2str(node.right) + ")"

        return resultStr
    }
}

File: Problems/0600_0699/0637_Average_of_Levels_in_Binary_Tree/Project_CS/OperateTreeNode.cs
using System;
using System.Collections.Generic;

public class OperateTreeNode
{
    public TreeNode CreateTreeNode(string flds)
    {
         return CreateSubTreeNode(flds.Split(","), 0, 0);
    }

    private TreeNode CreateSubTreeNode(string[] flds, int depth, int pos)
    {
        if (flds.Length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.Pow(2, i);
        
        if (cur_pos + pos > flds.Length - 1)
            return null;
        
        if (flds[cur_pos + pos] == "null")
            return null;

        if (flds[cur_pos + pos] == "")
            return null;

        try
        {
            TreeNode node = new TreeNode(int.Parse(flds[cur_pos + pos]));
            node.left = CreateSubTreeNode(flds, depth + 1, 2*pos);
            node.right = CreateSubTreeNode(flds, depth + 1, 2*pos + 1);

            return node;
        }
        catch (Exception e)
        {
            Console.WriteLine("\n" +  e.Message + "\n" +
                              "CreateSubTreeNode() Error ... flds[" + (cur_pos + pos).ToString() + "] = " + flds[cur_pos + pos] + "\n");
            Environment.Exit(-1);

            return null;
        }
    }

    List<string> resultList;

    public string TreeToStaircaseString(TreeNode node)
    {
        resultList = new List<string>();

        string outStr = TreeToStaircaseSubString(node, 0);
        resultList.Clear();

        return outStr;
    }

    private string TreeToStaircaseSubString(TreeNode node, int n)
    {
        if (node == null)
            return "";
        
        if (resultList.Count <= n)
            resultList.Add("(" + node.val.ToString() + ")");
        else
            resultList[n] += ",(" + node.val.ToString() + ")";

        if (node.left != null)
            TreeToStaircaseSubString(node.left, n + 1);
        if (node.right != null)
            TreeToStaircaseSubString(node.right, n + 1);

        return String.Join("\n", resultList) + "\n";
    }

    public string Tree2str(TreeNode node)
    {
        if (node == null)
            return "";

        string resultStr = node.val.ToString();

        if (node.left == null && node.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(node.left) + ")";
        if (node.right != null)
            resultStr += "(" + Tree2str(node.right) + ")";

        return resultStr;
    }
}
